virt_net: idempotency of create/stop actions

goneri opened the pull request:
Currently, if we try to stop or start a network two time in a row, the
second call will fail. With this patch:

- we don't recreate a network, if it exists
- we only stop a network if it's active, and so we avoid an exception
  saying the network is not active

##### SUMMARY

The following call:
```shell
ansible localhost -m virt_net -a "uri=qemu:///system command=start name=default" -b
```
Was raising the following error if called two times in a row:

```
localhost | FAILED! => {
    "changed": false,
    "msg": "Requested operation is not valid: network is already active"
}
```
##### ISSUE TYPE

Fixes: #26486.

##### COMPONENT NAME

- virt_net

goneri committed:
virt_net: idempotency of create/stop actions

Currently, if we try to stop or start a network two time in a row, the
second call will fail. With this patch:

- we don't recreate a network, if it exists
- we only stop a network if it's active, and so we avoid an exception
  saying the network is not active

goneri committed:
test: mock libvirt

Do not depend on the libvirt package. Instead we mock the method that
we actually need

mattclay commented on test/units/modules/cloud/misc/test_virt_net.py:
```diff
@@ -0,0 +1,55 @@
+import mock
+import pytest
+
+libvirt = pytest.importorskip('libvirt')
```
Instead of depending on `libvirt` for the unit tests, have you considered mocking the import?

See https://github.com/ansible/ansible/pull/50850 for an example of how that can be done.

goneri replied:
Yes, definitely. But I believe this also reduces the value of the unit-test.

mattclay replied:
For testing with actual dependencies such as `libvirt` I recommend using integration tests.

Unit tests can rely on mocking the imports of dependencies, and are good for verifying the internal logic of modules and module_utils. Integration tests use the actual dependencies and are good for verifying they work correctly with modules.

This allows both unit and integration tests to test against stable public APIs, rather than being dependent on the internals of dependencies which are subject to change.

goneri replied:
Libvirt comes with a test driver ( https://libvirt.org/drvtest.html ) that allow one to test some basic scenarios without doing any actual changes on the system. It does not cover everything but it's already a good start. This is what I use here.

Functional tests are IMO complementary and come with their own limitations (hard to cover all the user-cases, interaction with the local system, extra layers of complexity on top of the code, etc).

goneri committed:
add integration tests for virt_net

The current test validate the idempotency of `define` and `create` actions.

goneri committed:
test: enable virt_net test on RedHat 7 and 8

Define the dependency lists for RedHat 7 and RedHat 8.

mattclay commented on test/integration/targets/virt_net/aliases:
```diff
@@ -0,0 +1,5 @@
+destructive
+shippable/posix/group1
```
Unless this test works under docker without privileged mode, remove this and use `unsupported`.

Tests that require privileged mode under docker are not run on Shippable.

goneri replied:
ok, thanks for the clarification.

mattclay replied:
Sorry, I gave you bad instructions.

I forgot that this would run on our RHEL VMs, so having it in `shippable/posix/group1` is appropriate.

goneri replied:
np, I will push an update. I understand I also need to remove `unsupported`.

mattclay commented on test/integration/targets/virt_net/aliases:
```diff
@@ -0,0 +1,5 @@
+shippable/posix/group1
+skip/freebsd
+skip/osx
+needs/privileged
```
Does this test even work under docker? If not, remove this.

goneri replied:
It does.

goneri committed:
ci: use the unsupported alias

We use unsupported, unless the test can be run in docker without the
privileged mode.

See: https://github.com/ansible/ansible/pull/53276#discussion_r263141870

goneri committed:
tests that require privileged mode are run in VM

So it's fine to use `shippable/posix/group1`.

See: https://github.com/ansible/ansible/pull/53276#discussion_r263153452

mattclay commented on lib/ansible/modules/cloud/misc/virt_net.py:
```diff
@@ -417,19 +417,24 @@ def set_autostart(self, entryid, state):
+            return self.conn.create(entryid)
+        except libvirt.libvirtError as e:
+            if e.get_error_code() == libvirt.VIR_ERR_NETWORK_EXIST:
+                return None
```
Did you intend to raise the original exception here when the condition is not met?

goneri replied:
hum, not really. I pushed a patch to fix that.

mattclay commented on test/integration/targets/virt_net/aliases:
```diff
@@ -0,0 +1,4 @@
+shippable/posix/group1
+skip/freebsd
+skip/osx
+needs/privileged
```
Add the `destructive` alias since this test installs and uninstalls packages.

mattclay replied:
The `destructive` alias is still needed.

mattclay commented on test/integration/targets/virt_net/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    - name: Remove libvirt packages
+      package:
+        name: "{{ virt_net_packages }}"
+        state: absent
```
Packages should only be removed if they were installed by the test.

I'm especially concerned about removing lxml as that is pre-installed in all of our [distro test containers](https://github.com/ansible/distro-test-containers/).

goneri replied:
Ok. Good point, I will address that.

mattclay commented on test/units/modules/cloud/misc/test_virt_net.py:
```diff
@@ -0,0 +1,27 @@
+try:
+    import mock
+except ImportError:
+    from unittest import mock
```
Use `from units.compat import mock` instead.

goneri replied:
Oh, indeed. Thanks.

goneri committed:
virt_net/create raise unexpected libvirt exception

Continue to raise an exception during the `create()` if it's not
a `VIR_ERR_NETWORK_EXIST`.

goneri committed:
import mock from units.compat

goneri committed:
virt_net: do not call create() on "active" network

If we try to create() an active network on RHEL7/RHEL8, we get the
following exception:
  code: VIR_ERR_OPERATION_INVALID.
  msg: Requested operation is not valid: network is already active

We now ensure first that the network is NOT active.

goneri committed:
virt_net func test: only clean up the libvirt packages

To avoid any side-effect with some other tests, we only remove the libvirt
packages.

goneri committed:
test: virt_net: don't use assert_called()

`assert_called()` is python36+, we use `assert_called_with()` instead.

Akasurde commented:
@mattclay Could you please re-review this ? Thanks

goneri committed:
virt_net: add the destructive alias

Use the destructive alias since this test installs and uninstalls packages.

goneri commented:
Oh, done. Sorry for that.

Matt Clay <notifications@github.com> writes:

> mattclay requested changes on this pull request.
>
>
>
>> @@ -0,0 +1,4 @@
> +shippable/posix/group1
> +skip/freebsd
> +skip/osx
> +needs/privileged
>
> The `destructive` alias is still needed.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/ansible/ansible/pull/53276#pullrequestreview-215866625

-- 
    Gonéri Le Bouder

mattclay commented on test/integration/targets/virt_net/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    - name: Remove only the libvirt packages
+      package:
+        name: "{{ virt_net_packages|select('match', '.*libvirt.*')|list }}"
+        state: absent
```
Does removing the libvirt packages also remove the network configuration created earlier? If not, it should also be removed as part of the test -- which would also cover more module functionality in the tests as well.

goneri replied:
The service is stopped automatically when the packages are removed. But yes, it's a good idea to validate the removal of the network with virt_net. I will do that.

mattclay commented on test/integration/targets/virt_net/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    - name: Remove libvirt packages
+      package:
+        name: "{{ virt_net_packages }}"
+        state: absent
```
This remove is unnecessary with the always block below.

mattclay commented on test/integration/targets/virt_net/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    - name: Ensure the second call returns "unchanged"
+      assert:
+        that:
+          - "second_virt_net_start is not changed"
```
You can reduce test output by combining the two asserts.

goneri replied:
ok

mattclay commented on test/integration/targets/virt_net/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+    - "default.yml"
+
+- block:
+    - include_tasks: 'prepare.yml'
```
Why are these tasks in a separate file?

goneri replied:
It was to isolate the installation from the tests but Yeah, it's just two tasks. I merge the two files.

mattclay commented on test/units/modules/cloud/misc/conftest.py:
```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
```
This conftest will affect all tests in the same directory. The contents should probably be moved into the test_virt_net.py file instead, unless there is a chance of them being reused with other virt tests. If so, this file and the tests should be placed in a virt subdirectory.

goneri replied:
I move the files but none of the fixtures of the conftest.py have `autouse=True`. So they won't affect the behavior of the other tests.

goneri committed:
move the test in virt_net dir

This to isolate the scope of the conftest.py.

goneri committed:
test/virt_net: clean up the network at the end

Remove the network at the end of the installation and validate the
idempotency.
